File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/base/UaTransaction.java
package org.opcfoundation.webserver.service.transactions.base;

import org.opcfoundation.webapi.service.types.ServiceContext;

import java.util.concurrent.CompletableFuture;

public abstract class UaTransaction<T_Request, T_Response> {
    protected final ServiceContext serviceContext;
    private final int handleId;

    public UaTransaction(
            ServiceContext serviceContext,
            int handleId)
    {
        this.serviceContext = serviceContext;
        this.handleId = handleId;
    }

    public ServiceContext getServiceContext() {
        return serviceContext;
    }

    public Integer getHandleId() {
        return handleId;
    }

    public abstract T_Request getItem();

    public abstract T_Response getResult();

    public abstract CompletableFuture<Void> execute();
}

File: opcua-webservice-node/src/server/service/transactions/base/UaTransaction.ts
import { ServiceContext } from "../../..";

export abstract class UaTransaction<TRequest, TResponse> {
    protected readonly _serviceContext: ServiceContext;
    private readonly _handleId: number;

    constructor(serviceContext: ServiceContext, handleId: number) {
        this._serviceContext = serviceContext;
        this._handleId = handleId;
    }

    get serviceContext(): ServiceContext {
        return this.serviceContext;
    }

    get handleId(): number {
        return this._handleId;
    }

    abstract getItem(): TRequest;

    abstract getResult(): TResponse;

    abstract execute(): Promise<void>;
}